### ISSUE TITLE
Support conditional compilation / static if

### ISSUE THREAD

**Issue by bsansouci** [facebook/reason#1359]
Hey everyone :D 
I saw that Bucklescript has support for [condition compilation](https://bucklescript.github.io/bucklescript/Manual.html#_conditional_compilation_support_static_if) but it's a preprocessor fed into `-pp`, which means we can't use it in conjuction with Reason (since refmt is also a preprocessor).

How could we support static if statement like Bucklescript does? 
It would help the cross platform story a lot in my opinion!

**Comment by bobzhang** [comment]
note bucklescript support of conditional compilation is baked in, the `-pp` is a fallback when you use a normal ocaml compiler. `ppx` is overkill, it would be nice that reason would have it baked in but it is more challenging to the printer

**Comment by IwanKaramazow** [comment]
Bucklescript solves this very elegantly in lexing.
I'm not 100% sure how we should/can do this in Reason. Since the printer takes an ast & formats based on the tree, we'll need to augment the ast in some kind of way.

**Comment by bsansouci** [comment]
@hcarty's suggestion seems like the easiest and sanest for now

**Comment by bsansouci** [comment]
@IwanKaramazow @jordwalke It seems like refmt doesn't support the syntax for `https://github.com/janestreet/ppx_optcomp`. Do you know how we'd write that syntax https://github.com/mcclure/ppx_const in Reason? I couldn't find something that refmts.

**Comment by yyc-git** [comment]
we need "Support conditional compilation / static if" in reason which is baked in! 

@chenglou  could you have a plan to support it?

**Comment by yyc-git** [comment]
@bsansouci  how to use https://github.com/mcclure/ppx_const in reason? could you give me suggestion?

**Comment by bsansouci** [comment]
@yyc-git Hey, right now the only way to use ppx_const would be to depend on it on opam.
You need to run
```
opam switch 4.02.3+buckle-master
eval `opam config env`
```

to get the right version of the compiler

then add  `"ocamlfind-dependencies": ["ppx_const"]`  to your bsconfig.json

I'm not 100% sure it'll work though, I have not tried it.

**Comment by yyc-git** [comment]
@bsansouci Thanks for your reply!

**Comment by bsansouci** [comment]
@yyc-git Hey I just landed a change in bsb-native that should allow you to use [matchenv](https://github.com/bsansouci/matchenv). See https://github.com/Schmavery/reprocessing/blob/master/src/Reprocessing_Hotreload.re for the syntax and all.
You just need one file that'll be the one that will have its implementation swapped, and then call `include [%matchenv ...` like this, and switch to get the different implementation. `BSB_BACKEND` is a magical value that'll be filed in by bsb-native when compiling to each target :)

**Comment by yyc-git** [comment]
@bsansouci  Thankes a lot.

Is Bsb(js backend) support it? I need it in js backbend anyway.

**Comment by bsansouci** [comment]
Yes, bsb-native is 100% compatible with bsb. Just small extensions to support compiling to bytecode/native :)

**Comment by yyc-git** [comment]
@bsansouci  Good job~

now I'm using Reason 3, which require bsb 2.0. so does bsb-native support reason3? support bsb 2.0?

**Comment by bsansouci** [comment]
Yup bsb-Native supports reason 3 :)
